packages/server-renderer/src/render.ts

```diff
@@ -94,22 +94,26 @@ export function renderComponentVNode(
   const instance = createComponentInstance(vnode, parentComponent, null)
   const res = setupComponent(instance, true /* isSSR */)
   const hasAsyncSetup = isPromise(res)
-  const prefetches = instance.sp /* LifecycleHooks.SERVER_PREFETCH */
-  if (hasAsyncSetup || prefetches) {
-    let p: Promise<unknown> = hasAsyncSetup
+  const getPrefetches = () => instance.sp /* LifecycleHooks.SERVER_PREFETCH */
+  if (hasAsyncSetup || getPrefetches()) {
+    const setupResolved: Promise<unknown> = hasAsyncSetup
       ? (res as Promise<void>)
       : Promise.resolve()
-    if (prefetches) {
-      p = p
-        .then(() =>
-          Promise.all(
+    const setupAndPrefetched = setupResolved
+      .then(() => {
+        // instance.sp may be null until an async setup resolves, so evaluate it here
+        const prefetches = getPrefetches()
```

Comment:
It seems `getPrefetches` will be called twice, if `hasAsyncSetup` equals `false`.

Reply:
Reading `instance.sp` only once and caching that value to be used after set up resolves is what caused the bug.

What is the concern about accessing `instance.sp` twice?

Reply:
@edison1105 I pushed a commit which:

* Makes `renderComponentVNode` async. It now always returns a promise.
* Uses `await` directly on the `setupComponent` call. No need to inspect it to see if it is a promise.
* Simplifies the implementation to remove branching logic.
* Reads `instance.sp` once, after setup.

Please let me know if this is what you had in mind. This extra commit is more involved.

Reply:
I'm not sure if it's reasonable to change `renderComponentVNode` to async. see https://github.com/vuejs/core/pull/11340#issue-2405010707
Let's wait for reviews from other team members

Reply:
@edison1105 Ok, I see that there is deliberate synchronous fast paths in place. In that case, I'm happy to revert the last commit or do whatever is needed here.

Reply:
What do you think of changing it like this?
```ts
  let prefetches = instance.sp /* LifecycleHooks.SERVER_PREFETCH */
  if (hasAsyncSetup || prefetches) {
    let p: Promise<unknown> = (
      hasAsyncSetup
        // instance.sp may be null until an async setup resolves, so evaluate it here
        ? (res as Promise<void>).then(() => (prefetches = instance.sp))
        : Promise.resolve()
    )
      .then(() => {
        if (prefetches) {
          return Promise.all(
            prefetches.map(prefetch => prefetch.call(instance.proxy)),
          )
        }
      })
      // Note: error display is already done by the wrapped lifecycle hook function.
      .catch(NOOP)
    return p.then(() => renderComponentSubTree(instance, slotScopeId))
  } else {
    return renderComponentSubTree(instance, slotScopeId)
  }
```

Reply:
@edison1105 I pushed up that change in 7af4f77182b3838fac07acc994ec3f846570a02d, and then I went a little further and optimized away two promises in e159b16ef3dc65de9a1c5f319d59e06d89547bc9.

1. Removes the unnecessary `hasAsyncSetup` ternary conditional by using `Promise.resolve` directly around the `res` object because it returns the [exact same promise if it is given one](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve).
2. Removes a new promise from by removing a `.then` callback
3. Removes another new promise by removing a `.catch` callback in favor of assigning the noop as the second argument

So maybe this will be a little faster or at least produce a little less garbage.